### PR TITLE
fix: return a friendly error if the dialer is closed

### LINKED ISSUE #522: Occasional "context cancelled" dial errors even though we don't use cancellable contexts
### Bug Description

We use an AlloyDB cluster from a Go service in Cloud Run, and it mostly works fine, except that sometimes we get errors such as:

> failed to connect to `host=/tmp user=foo@project-bar.iam database=foo`: dial error (context canceled)

Which doesn't make sense, given that the database connection pool (https://pkg.go.dev/github.com/jackc/pgx/v5/pgxpool) is opened with a background context, and the SQL statement is executed with a TODO context, and neither are cancellable.

The Go program in Cloud Run is connected to AlloyDB via a VPC connector, as outlined in https://cloud.google.com/alloydb/docs/quickstart/integrate-cloud-run.

We use the following versions:

* go 1.22.0
* cloud.google.com/go/alloydbconn v1.7.0 
* github.com/jackc/pgx/v5 v5.5.3

We set up pgxpool with the alloydb connector as best we could, looking at the examples in your documentation. The relevant bits of the code are below; hopefully we didn't make any mistakes.

### Example code (or command)

```Go
// opening the database

ctx := context.Background()
config, err := pgxpool.ParseConfig(dbURL)
if err != nil {
	return nil, err
}
d, err := alloydbconn.NewDialer(ctx, alloydbconn.WithIAMAuthN())
if err != nil {
	log.Fatalf("failed to initialize dialer: %v", err)
}
defer d.Close()
config.ConnConfig.DialFunc = func(ctx context.Context, _ string, instance string) (net.Conn, error) {
	return d.Dial(ctx, cfg.AlloyDBInstance)
}
pool, err := pgxpool.NewWithConfig(ctx, config)
if err != nil {
	return nil, err
}
srv.db = pool

// using the db pool to run the statement above which failed only once

ctx := context.TODO()
_, err := srv.db.Exec(ctx, `
	UPDATE repos_cache
	SET updated = '2000-01-01 00:00+00'::timestamptz,
		updated_collaborators = '2000-01-01 00:00+00'::timestamptz
	WHERE provider = $1
	AND owner_name = $2
	AND repo_name = $3
`, repo.provider, repo.ownerName, repo.repoName)
if err != nil {
	return err
}
```

### Stacktrace

_No response_

### Steps to reproduce?

We aren't able to reliably reproduce this error. It happens occasionally in our live service; we have seen the error twice in the past two weeks.

### Environment

1. OS type and version: Linux on Cloud Run, unclear what version it is
2. Go version: 1.22.0
3. AlloyDB Go Connector version: v1.7.0

### Additional Details

The code isn't in a public repo, apologies. I'm more than happy to try to provide any more information that is requested.

### PR DESCRIPTION
**Pull request by enocom**
If the dialer has already been closed, return a clear error.

Fixes #522

### CODE REVIEW

File: dialer.go [hunk 1]
```diff
@@ -56,6 +56,12 @@ const (
 )
 
 var (
+	// ErrDialerClosed is used when a caller invokes Dial after closing the
+	// Dialer.
+	ErrDialerClosed = errors.New(
+		"Dialer has been closed. Close should be " +
+			"called only when a database connection is no longer needed.",
```
Comment (mvdan) [inline dialer.go:63]: usually errors aren't punctuated or capitalized, as they are chained together as a single sentence. for example, from std:

```
$ git grep -i 'errclosed.*errors\.new'
compress/lzw/reader.go:var errClosed = errors.New("lzw: reader/writer is closed")
database/sql/fakedb_test.go:var errClosed = errors.New("fakedb: statement has been closed")
internal/oserror/errors.go: ErrClosed     = errors.New("file already closed")
io/pipe.go:var ErrClosedPipe = errors.New("io: read/write on closed pipe")
net/http/h2_bundle.go:  http2errClosedPipeWrite        = errors.New("write on closed buffer")
net/http/h2_bundle.go:  http2errClosedBody         = errors.New("body closed by handler")
net/http/h2_bundle.go:var http2errClosedResponseBody = errors.New("http2: response body closed")
net/http/httputil/persist.go:var errClosed = errors.New("i/o operation on closed connection")
net/http/roundtrip_js.go:var errClosed = errors.New("net/http: reader is closed")
```

How about `alloydb: dialer is closed`? Given how almost every other close-able resource cannot be used once closed (files, connections, clients, etc), I don't think you need to explain when or when not to call close in the error message.

**Review by mvdan — COMMENTED** [review]
Thanks, SGTM! Just a thought about the error message itself.

File: dialer.go (hunk 1 above)
Comment (enocom) [reply dialer.go:63]: I should know better :sweat_smile: Thanks for the suggestion. Will get that fixed.

**Review by mvdan — APPROVED** [review]

**Review by jackwotherspoon — APPROVED** [review]
LGTM ✅

**Pull request merged**